src/compiler/program.ts

```diff
@@ -2275,7 +2275,7 @@ namespace ts {
                     programDiagnostics.add(createFileDiagnostic(firstNonExternalModuleSourceFile, span.start, span.length, Diagnostics.Cannot_compile_namespaces_when_the_isolatedModules_flag_is_provided));
                 }
             }
-            else if (firstExternalModuleSourceFile && languageVersion < ScriptTarget.ES6 && options.module === ModuleKind.None) {
+            else if (firstExternalModuleSourceFile && !isDeclarationFile(firstExternalModuleSourceFile) && languageVersion < ScriptTarget.ES6 && options.module === ModuleKind.None) {

```

Comment:
This is not enough because for program with a.d.ts and b.ts the error wont be reported in b.ts
You technically want firstNonAmbientExternalModuleSourceFile here